src/Renci.SshNet/Common/PipeInputStream.cs

```diff
@@ -0,0 +1,112 @@
+﻿namespace Renci.SshNet.Common 
+{
+    using System;
+    using System.IO;
+
+    internal class PipeInputStream : Stream
+    {
+        private LinkedListQueue<byte[]> _queue;
+        private byte[] _current;
+        private int _currentPosition;
+        private bool _isDisposed;
+
+        public PipeInputStream(LinkedListQueue<byte[]> queue)
+        {
+            _queue = queue;
+        }
+
+        public override void Flush()
+        {
+        }
+
+        public override long Seek(long offset, SeekOrigin origin)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override void SetLength(long value)
+        {
+            throw new NotSupportedException();
+        }
+
+        public override int Read(byte[] buffer, int offset, int count)
+        {
+            if (buffer == null)
+                throw new ArgumentNullException("buffer");
+            if (offset + count > buffer.Length)
+                throw new ArgumentException("The sum of offset and count is greater than the buffer length.");
+            if (offset < 0 || count < 0)
+                throw new ArgumentOutOfRangeException("offset", "offset or count is negative.");
+            if (_isDisposed)
+                throw CreateObjectDisposedException();
+
+            if (_current == null || _currentPosition == _current.Length)
+            {
+                if (_queue.IsCompleted || !_queue.TryTake(out _current))
+                    return 0;
+
+                _currentPosition = 0;
+            }
+
+            var avail = _current.Length - _currentPosition;
```

Comment:
It's still up to Stream specification as it blocks until *some* bytes are available and it will return 0 on EOS. Can be reverted back to a while() loop if it would be better in your opinion to return the full requested buffer size until EOS.